src/daemon/daemon.cpp

```diff
@@ -339,9 +339,12 @@ bool Daemon::deactivate(bool emitSignals) {
     }
   }
 
-  if (emitSignals) {
-    emit disconnected();
-  }
+  // Emit signals upon return.
+  auto guard = qScopeGuard([&]() {
```

Comment:
Why do we need this, out of curiosity?

Reply:
This was added for the wasm, which is a single-threaded platform and therefore the signal handlers that run in response to the `disconnected()` signal are just called directly. This results in the controller trying to do stuff before the `deactivate()` call even finishes and that caused all kinds of chaos.

Normally we don't see this kind of thing because the messages to the daemon are queued through an IPC pipe of some sort.

I suppose this should be considered a bug, and the controller ought to be prepared for a `disconnected()` signal even in the middle of `deactivate()` call, but this workaround was easier in the heat of the moment.